### PR TITLE
Pending transactions fix

### PR DESCRIPTION
**Pull request by dyukorev**
Android fix: Pass a callback pointer to the native code to avoid the issue with multiple calls product list function.
Android and iOs: Add the "process_pending_transactions" function which can be called to process started but not completed transactions (before they was pending forever and was blocking buying the same product again).
Fix is pretty massive sorry for that.

### CODE REVIEW

File: extension-iap/src/iap_ios.mm [hunk 1]
```diff
@@ -324,28 +325,25 @@ static void HandlePurchaseResult(IAPCommand* cmd)
     assert(top == lua_gettop(L));
 }
 
-@implementation SKPaymentTransactionObserver
-    - (void)paymentQueue:(SKPaymentQueue *)queue updatedTransactions:(NSArray *)transactions
-    {
-        for (SKPaymentTransaction* transaction in transactions) {
-
-            if ((!self.m_IAP->m_AutoFinishTransactions) && (transaction.transactionState == SKPaymentTransactionStatePurchased)) {
+static void processTransactions(IAP* m_IAP, NSArray* transactions) {
```
Comment (JCash) [inline extension-iap/src/iap_ios.mm]: nitpick: function arguments should be lowercase, and also not have the prefix.
Replace with `iap` here.

File: extension-iap/src/iap_ios.mm [hunk 2]
```diff
@@ -477,18 +487,6 @@ static int IAP_SetListener(lua_State* L)
...
-        observer.m_IAP = &g_IAP;
-        // NOTE: We add the listener *after* a lua listener is set
-        // The payment queue is persistent and "old" transaction might be processed
-        // from previous session. We call "finishTransaction" when appropriate
-        // for all transaction and we must ensure that the result is delivered to lua.
-        [[SKPaymentQueue defaultQueue] addTransactionObserver: observer];
-        g_IAP.m_Observer = observer;
-    }
-
     return 0;
```
Comment (JCash) [inline extension-iap/src/iap_ios.mm]: This is a behavioral change.
Previously, when when setting the listener, you could receive callbacks.
With this change, it won't happen? That might potentially break our users code I assume.

Please motivate the change, and also if the change should be considered a major number bump.

File: extension-iap/src/iap_ios.mm [hunk 3]
```diff
@@ -506,6 +504,7 @@ static int IAP_GetProviderId(lua_State* L)
     {"restore", IAP_Restore},
     {"set_listener", IAP_SetListener},
     {"get_provider_id", IAP_GetProviderId},
+    {"process_pending_transactions", IAP_ProcessPendingTransactions},
```
Comment (JCash) [inline extension-iap/src/iap_ios.mm]: Seeing as you've added this function, I assume you intended a major version change, but forgot to flag for it in the description?

File: extension-iap/src/iap_ios.mm [hunk 4]
```diff
@@ -557,6 +563,9 @@ static void IAP_OnCommand(IAPCommand* cmd, void*)
 static dmExtension::Result UpdateIAP(dmExtension::Params* params)
 {
     IAP_Queue_Flush(&g_IAP.m_CommandQueue, IAP_OnCommand, 0);
+    if (g_IAP.m_Observer) {
+        IAP_Queue_Flush(&g_IAP.m_ObservableQueue, IAP_OnCommand, 0);
```
Comment (JCash) [inline extension-iap/src/iap_ios.mm]: This extra queue isn't needed?
What's the rationale behind it?
I would assume pushing the commands to the `m_CommandQueue` would suffice.

File: extension-iap/src/iap_ios.mm (hunk 2 above)
Comment (britzl) [reply extension-iap/src/iap_ios.mm]: Exactly. You'd get callbacks for any pending purchases when you set the listener. The same behaviour applies to both iOS and Android. I don't think we should change this unless there is a very compelling reason.

File: extension-iap/src/iap_ios.mm (hunk 4 above)
Comment (dyukorev) [reply extension-iap/src/iap_ios.mm]: There is another way to solve this: put to the queue command with empty callback and check if its empty on queue processing, and when listener setup iterate queue and setup callback for empty callbacks but I think its worse.

File: extension-iap/src/iap_ios.mm (hunk 3 above)
Comment (AGulev) [reply extension-iap/src/iap_ios.mm]: Could you please add documentation for the new method as well?

File: extension-iap/src/iap_ios.mm (hunk 2 above)
Comment (JCash) [reply extension-iap/src/iap_ios.mm]: I see, and I agree with you that it makes sense to do it this way.
Thanks for the explanation.

File: extension-iap/src/iap_ios.mm (hunk 3 above)
Comment (JCash) [reply extension-iap/src/iap_ios.mm]: Agreed, please add documentaiton.

File: extension-iap/src/iap_ios.mm (hunk 4 above)
Comment (JCash) [reply extension-iap/src/iap_ios.mm]: Now that I understand what this queue is for, and how the code flow is, I think this PR is a good way of solving it.

File: extension-iap/src/iap_ios.mm (hunk 3 above)
Comment (britzl) [reply extension-iap/src/iap_ios.mm]: @dyukorev Could you please address this?

File: extension-iap/src/iap_ios.mm (hunk 1 above)
Comment (britzl) [reply extension-iap/src/iap_ios.mm]: @dyukorev Could you please address this?

**Comment by britzl** [comment]
@dyukorev We are ready to make a final review and then merge this once the requested changes have been made.

**Pull request merged**